crates/client/src/lib.rs

```diff
@@ -1,14 +1,183 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use std::collections::HashMap;
+use zbus::zvariant::Value;
+
+pub struct HintsData {
+    pub urgency: Option<String>,
+    pub category: Option<String>,
+    pub desktop_entry: Option<String>,
+    pub image_path: Option<String>,
+    pub resident: Option<bool>,
+    pub sound_file: Option<String>,
+    pub sound_name: Option<String>,
+    pub suppress_sound: Option<bool>,
+    pub transient: Option<bool>,
+    pub action_icons: Option<bool>,
+}
+
+pub struct NotiClient<'a> {
+    dbus_client: dbus::client::Client<'a>,
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+impl<'a> NotiClient<'a> {
+    pub async fn init() -> anyhow::Result<Self> {
+        let client = dbus::client::Client::init().await?;
+        Ok(Self {
+            dbus_client: client,
+        })
+    }
+
+    pub async fn send_notification(
+        &self,
+        id: u32,
+        app_name: &'a str,
+        icon: &'a str,
+        summary: &'a str,
+        body: &'a str,
+        timeout: i32,
+        actions: &'a Vec<String>,
+        hints: &'a Vec<String>,
+        hints_data: &'a HintsData,
+    ) -> anyhow::Result<()> {
+        let new_hints = Self::build_hints(&hints, &hints_data)?;
+        let actions = Self::build_actions(actions)?;
+
+        self.dbus_client
+            .notify(
+                &app_name, id, &icon, &summary, &body, actions, new_hints, timeout,
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn get_server_info(&self) -> anyhow::Result<()> {
+        let server_info = self.dbus_client.get_server_information().await?;
+        println!(
+            "Name: {}\nVendor: {}\nVersion: {}\nSpecification version: {}",
+            server_info.0, server_info.1, server_info.2, server_info.3
+        );
+
+        Ok(())
+    }
+
+    fn build_actions(actions: &'a [String]) -> anyhow::Result<Vec<&'a str>> {
+        let mut new_actions = Vec::with_capacity(actions.len() * 2);
+
+        for entry in actions {
+            if let Some((action_name, action_desc)) = entry.split_once(':') {
+                new_actions.push(action_name.trim());
+                new_actions.push(action_desc.trim());
+            }
+        }
+
+        Ok(new_actions)
+    }
+
+    fn build_hints(
+        hints: &'a [String],
+        hints_data: &'a HintsData,
+    ) -> anyhow::Result<HashMap<&'a str, Value<'a>>> {
+        let mut hints_map: HashMap<&str, Value> = HashMap::with_capacity(hints.len());
+
+        for entry in hints {
+            let parts: Vec<&'a str> = entry.split(':').collect();
+
+            if parts.len() == 3 {
+                let hint_type = parts[0].trim();
+                let hint_name = parts[1].trim();
+                let hint_value = parts[2].trim();
+
+                let value = Self::parse_hint_value(hint_type, hint_value)?;
+                hints_map.insert(hint_name, value);
+            }
```

Comment:
Same here, instead of this use `if parts.len() != 3 { bail("according message") }` because user may do incorrect input and need to tell immediate where he did wrong.